crates/sparrow-execution/src/lib.rs

```diff
@@ -8,254 +8,193 @@
+impl PlanExecutor {
+    /// Create the plan executor for the given plan.
+    ///
+    /// This creates the worker threads but does not inject inputs, meaning
+    /// the threads won't have anything to do until `run` is called.
+    pub fn try_new(
         query_id: String,
-        input_type: DataType,
-        mut input: tokio::sync::mpsc::Receiver<Batch>,
-        output_type: DataType,
+        plan: sparrow_physical::Plan,
+        mut inputs: HashMap<Uuid, tokio::sync::mpsc::Receiver<Batch>>,
         output: tokio::sync::mpsc::Sender<Batch>,
-    ) -> error_stack::Result<(), Error> {
-        let mut worker_pool = WorkerPool::start(query_id).change_context(Error::Creating)?;
+    ) -> error_stack::Result<Self, Error> {
+        let mut executor = PlanExecutor {
+            worker_pool: WorkerPool::new(query_id).change_context(Error::Creating)?,
+            source_tasks: SourceTasks::default(),
+        };
 
-        // This sets up some fake stuff:
-        // - We don't have sources / sinks yet, so we use tokio channels.
-        // - We create a "hypothetical" scan step (0)
-        // - We create a hard-coded "project" step (1)
-        // - We output the results to the channel.
+        let sink_pipeline = executor
+            .worker_pool
+            .add_pipeline(1, WriteChannelPipeline::new(output));
+
+        // Map from the producing step ID to the consumers.
+        let mut consumers: HashMap<StepId, PipelineInputs> = HashMap::new();
+        consumers
+            .entry(plan.steps.last_idx())
+            .or_default()
+            .add_consumer(sink_pipeline, 0);
+
+        // Iterate in reverse so the receivers (consumers) are created before the senders (producers).
+        for pipeline in plan.pipelines.iter().rev() {
+            let first_step_id = *pipeline.steps.first().expect("at least one step");
+            let last_step_id = *pipeline.steps.last().expect("at least one step");
+            let sink = consumers
+                .remove(&last_step_id)
+                .expect("at least one consumer for step");
+
+            // This is a limitation of how the `PipelineInput` is identified. Ideally,
+            // any pipeline should be connect-able to multiple down-stream pipelines.
+            debug_assert_eq!(sink.len(), 1, "Only one consumer currently supported");
+
+            let first_step_inputs = &plan.steps[first_step_id].inputs;
+
+            // Create a transform pipeline (if possible), othrewise create the
+            // appropriate non-transform pipeline.
+            let pipeline = if pipeline
+                .steps
+                .iter()
+                .all(|id| plan.steps[*id].kind.is_transform())
+            {
+                debug_assert_eq!(
+                    first_step_inputs.len(),
+                    1,
+                    "Transforms should have a single input"
+                );
+                let first_step_input_id = first_step_inputs[0];
```

Comment:
Could have `add_transform_pipeline` determine this -- would keep the main code simpler.